src/connectors/query-rules/__tests__/connectQueryRules-test.ts

```diff
@@ -0,0 +1,450 @@
+    });
+
+    test('throws with a non-functon tracked filter', () => {
+      expect(() => {
+        makeWidget({
+          // @ts-ignore
+          trackedFilters: {
+            brand: ['Samsung'],
+          },
+        });
+      }).toThrowErrorMatchingInlineSnapshot(`
+"'The \\"brand\\" filter value in the \`trackedFilters\` option expects a function.
+
+See documentation: https://www.algolia.com/doc/api-reference/widgets/query-rules/js/#connector"
+`);
+    });
+  });
+
+  describe('lifecycle', () => {
+    test('calls the render function on init', () => {
+      const helper = createFakeHelper();
+      const widget = makeWidget({});
+
+      widget.init({ helper, state: helper.state, instantSearchInstance: {} });
+
+      const [renderingParameters, isFirstRender] = renderFn.mock.calls[0];
+
+      expect(isFirstRender).toBe(true);
+
+      const {
+        userData,
+        instantSearchInstance,
+        widgetParams,
+      } = renderingParameters;
+
+      expect(userData).toEqual([]);
+      expect(instantSearchInstance).toEqual({});
+      expect(widgetParams).toEqual({});
+    });
+
+    test('calls the render function on render', () => {
+      const helper = createFakeHelper();
+      const widget = makeWidget({});
+
+      widget.init({ helper, state: helper.state, instantSearchInstance: {} });
+      widget.render({
+        helper,
+        results: new SearchResults(helper.getState(), [{ hits: [] }]),
+        instantSearchInstance: {},
+      });
+
+      {
+        const [renderingParameters, isFirstRender] = renderFn.mock.calls[1];
+
+        expect(isFirstRender).toBe(false);
+
+        const {
+          userData,
+          instantSearchInstance,
+          widgetParams,
+        } = renderingParameters;
+
+        expect(userData).toEqual([]);
+        expect(instantSearchInstance).toEqual({});
+        expect(widgetParams).toEqual({});
+      }
+
+      widget.render({
+        helper,
+        results: new SearchResults(helper.getState(), [
+          { hits: [], userData: [{ banner: 'image.png' }] },
+        ]),
+        instantSearchInstance: {},
+      });
+
+      {
+        const [renderingParameters, isFirstRender] = renderFn.mock.calls[2];
+
+        expect(isFirstRender).toBe(false);
+
+        const {
+          userData,
+          instantSearchInstance,
+          widgetParams,
+        } = renderingParameters;
+
+        expect(userData).toEqual([{ banner: 'image.png' }]);
+        expect(instantSearchInstance).toEqual({});
+        expect(widgetParams).toEqual({});
+      }
+    });
+
+    test('calls the unmount function on dispose', () => {
+      const helper = createFakeHelper();
+      const widget = makeWidget({});
+
+      widget.init({ helper, state: helper.state, instantSearchInstance: {} });
+      widget.dispose({ state: helper.getState() });
+
+      expect(unmountFn).toHaveBeenCalledTimes(1);
+    });
+
+    test('calls the unmount function on dispose with trackedFilters', () => {
```

Comment:
The reason why I added this test was to make sure that `unmount` is called in `dispose` in all cases because we return the state early.

For example, the test would fail if the implementation was like that:

```js
dispose({ state }) {
  if (Object.keys(trackedFilters).length > 0) {
    const reinitRuleContexts = state
      .getQueryParameter('ruleContexts')
      .filter((rule: string) => !addedRuleContexts.includes(rule));

    return state.setQueryParameter('ruleContexts', reinitRuleContexts);
  }

  // Does not get called because `trackedFilters` was defined.
  unmount();

  return state;
},
```

The test doesn't provide much value in itself. I wrote it because when refactoring the `dispose` function I saw some potential misbehaviors here.

I can remove it if it feels unnecessary.

Reply:
Ok, maybe a cleare test name would help to understand why it’s there.